tests/test_crf_cpu.py

```diff
@@ -0,0 +1,450 @@
+                # Class 1
+                [
+                    # Slice 0
+                    [
+                        [0.000000, 0.000000, 0.000000, 0.000192, 0.278878],
+                        [0.000000, 0.000000, 0.000000, 0.000242, 0.333975],
+                        [0.000000, 0.000000, 0.000021, 0.004106, 0.422541],
+                        [0.000213, 0.000261, 0.004275, 0.065830, 0.511296],
+                        [0.308355, 0.358972, 0.439873, 0.518282, 0.568820],
+                    ],
+                    # Slice 1
+                    [
+                        [0.000000, 0.000000, 0.000000, 0.000257, 0.349584],
+                        [0.000000, 0.000000, 0.000001, 0.007253, 0.891966],
+                        [0.000000, 0.000001, 0.001459, 0.597630, 0.992878],
+                        [0.000289, 0.007891, 0.608059, 0.996642, 0.999560],
+                        [0.384477, 0.902880, 0.993401, 0.999573, 0.999635],
+                    ],
+                    # Slice 2
+                    [
+                        [0.000000, 0.000000, 0.000025, 0.004759, 0.456878],
+                        [0.000000, 0.000002, 0.001606, 0.618019, 0.993414],
+                        [0.000027, 0.001687, 0.629762, 0.999404, 0.999966],
+                        [0.005389, 0.638683, 0.999427, 0.999999, 1.000000],
+                        [0.494608, 0.994138, 0.999968, 1.000000, 1.000000],
+                    ],
+                    # Slice 3
+                    [
+                        [0.000308, 0.000361, 0.005636, 0.080287, 0.553317],
+                        [0.000374, 0.009877, 0.652810, 0.997105, 0.999610],
+                        [0.006128, 0.663335, 0.999475, 0.999999, 1.000000],
+                        [0.089296, 0.997324, 0.999999, 1.000000, 1.000000],
+                        [0.586036, 0.999646, 1.000000, 1.000000, 1.000000],
+                    ],
+                    # Slice 4
+                    [
+                        [0.425504, 0.466694, 0.530665, 0.580554, 0.609597],
+                        [0.475907, 0.927820, 0.994913, 0.999646, 0.999682],
+                        [0.550638, 0.995125, 0.999972, 1.000000, 1.000000],
+                        [0.606569, 0.999669, 1.000000, 1.000000, 1.000000],
+                        [0.637583, 0.999705, 1.000000, 1.000000, 1.000000],
+                    ],
+                ],
+            ],
+        ],
+    ],
+]
+
+
+@skip_if_no_cpp_extension
+class CRFTestCaseCpu(unittest.TestCase):
+    @parameterized.expand(TEST_CASES)
+    def test(self, test_case_description, params, input, features, expected):
+
+        # Create input tensors
+        input_tensor = torch.from_numpy(np.array(input)).to(dtype=torch.float, device=torch.device("cpu"))
+        feature_tensor = torch.from_numpy(np.array(features)).to(dtype=torch.float, device=torch.device("cpu"))
+
+        # apply filter
+        crf = CRF(*params)
+        output = crf(input_tensor, feature_tensor).cpu().numpy()
```

Comment:
So how to use it in backprop? I'm trying to do this, but get some typeerrors (`TypeError: phl_filter(): incompatible function arguments. The following argument types are supported:
    1. (arg0: at::Tensor, arg1: at::Tensor) -> at::Tensor`)
```suggestion
        input_tensor.requires_grad = True
        input_tensor.retain_grad()
        output = crf(input_tensor, feature_tensor)
        output.sum().backward()
        output = output.cpu().numpy()
```

Reply:
A good point, I think the PHLFilter's back propergation is only implemented for one of the input tensors, which is an oversight. I'll get working on the fix, I can either add the changes to this or make a new PR?

Reply:
the current impl. is enough for creating this tutorial https://github.com/NifTK/NiftyNet/blob/dev/demos/crf_as_rnn/crf_as_rnn_inference_demo.ipynb shall we merge this first? but it should raise a NotImpelmentedError when we run backprop

Reply:
I had intended to move on to investigating the back propergation after it was in so that suits me. Ill make it raise an error for now and open a new PR for the fix.